internal/controller/management/projectconfigs/project_configs.go

```diff
@@ -0,0 +1,311 @@
+	"errors"
+	"fmt"
+	"path"
+	"time"
+
+	multierr "github.com/hashicorp/go-multierror"
+	"github.com/kelseyhightower/envconfig"
+	corev1 "k8s.io/api/core/v1"
+	kubeerr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/conditions"
+	"github.com/akuity/kargo/internal/controller"
+	"github.com/akuity/kargo/internal/kubeclient"
+	"github.com/akuity/kargo/internal/logging"
+	"github.com/akuity/kargo/internal/webhook/external"
+)
+
+type ReconcilerConfig struct {
+	ExternalWebhookServerBaseURL string `envconfig:"EXTERNAL_WEBHOOK_SERVER_BASE_URL"`
+	MaxConcurrentReconciles      int    `envconfig:"MAX_CONCURRENT_PROJECT_RECONCILES" default:"4"`
+}
+
+func ReconcilerConfigFromEnv() ReconcilerConfig {
+	cfg := ReconcilerConfig{}
+	envconfig.MustProcess("", &cfg)
+	return cfg
+}
+
+type reconciler struct {
+	cfg    ReconcilerConfig
+	client client.Client
+
+	syncWebhookReceiversFn func(
+		context.Context,
+		*kargoapi.ProjectConfig,
+	) ([]kargoapi.WebhookReceiver, error)
+}
+
+func SetupReconcilerWithManager(
+	ctx context.Context,
+	kargoMgr manager.Manager,
+	cfg ReconcilerConfig,
+) error {
+	_, err := ctrl.NewControllerManagedBy(kargoMgr).
+		For(&kargoapi.ProjectConfig{}).
+		WithEventFilter(
+			predicate.Funcs{
+				DeleteFunc: func(event.DeleteEvent) bool {
+					return false
+				},
+			},
+		).
+		WithOptions(controller.CommonOptions(cfg.MaxConcurrentReconciles)).
+		Build(newReconciler(kargoMgr.GetClient(), cfg))
+	if err != nil {
+		return fmt.Errorf("error creating Project reconciler: %w", err)
+	}
+
+	logging.LoggerFromContext(ctx).Info(
+		"Initialized ProjectConfig reconciler",
+	)
+	return nil
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	r := &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+	r.syncWebhookReceiversFn = r.syncWebhookReceivers
+	return r
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *reconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"project-config", req.NamespacedName.Name,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+
+	// Fetch the ProjectConfig instance
+	projectConfig := &kargoapi.ProjectConfig{}
+	if err := r.client.Get(ctx, req.NamespacedName, projectConfig); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if projectConfig.DeletionTimestamp != nil {
+		logger.Debug("ProjectConfig is being deleted; nothing to do")
+		return ctrl.Result{}, nil
+	}
```

Comment:
Even if they are technically not required at this point, it is probably good to set a finalizer.